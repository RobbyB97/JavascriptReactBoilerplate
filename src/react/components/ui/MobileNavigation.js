/* Packages */
import React from 'react';
import {NavLink} from 'react-router-dom';
import {connect} from 'react-redux';


/* Component */
export class MobileNavigation extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            navMenu_Mobile: false
        };
    };

    
    navMenu_Mobile__Toggle = () => {
        this.props.navMenu_Mobile__Toggle();
    };


    render() {
        return (
            <section id="MobileNavigation">
                <aside 
                    data-active={this.state.navMenu_Mobile}
                    className="MobileNavigation__Menu"
                >
                    <ul className="MobileNavigation__list">
                        {this.props.internal_links &&
                            this.props.internal_links.map((link) => (
                                <li key={link.text}>
                                    <a href={link.href}>
                                        {link.text}
                                    </a>
                                </li>
                            ))
                        }
                    </ul>
                </aside>
            </section>
        );
    };
};


/* Connect to store */
const mapStateToProps = (state) => {
    return {
        navMenu_Mobile: state.ui.navMenu_Mobile
    };
};


const mapDispatchToProps = (dispatch) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default MobileNavigation;