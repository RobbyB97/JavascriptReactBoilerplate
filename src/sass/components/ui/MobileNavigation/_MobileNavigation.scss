#MobileNavigation {
    @media #{$desktop_breakpoint} {
        display: none;
    }
}


.MobileNavigation__button {
    position: fixed;
    top: 12px;
    right: 12px;
    font-size: 2em;
    z-index: 8;
}


.MobileNavigation__menu {
    position: fixed;
    top: 0;
    right: 0;

    &[data-active=false] {
        opacity: 0;
        transition: 200ms;
        width: 0;
        height: 0;
    }

    &[data-active=true] {
        display: block;
        opacity: 1;
        transition: 200ms;
        background: desaturate(darken($purple, 35%), 70%);
        z-index: 7;
        width: 100%;
        height: 100%;
    }
}


.MobileNavigation__list {
    padding: 0;
    margin-top: 60px;
}


.MobileNavigation__link {
    list-style-type: none;
    margin: 1em auto;

    a {
        color: $white;
        font-size: 1.4em;
        text-decoration: none;
    }
}


.MobileNavigation__sublink_list {
    background: desaturate(darken($purple, 50%), 80%);
    list-style-type: none;
    padding: 0;
}


.MobileNavigation__sublink {

}