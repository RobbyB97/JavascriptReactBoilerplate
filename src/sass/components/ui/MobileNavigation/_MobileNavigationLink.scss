.MobileNavigationLink {
    list-style-type: none;
    padding: 0;
    background: rgba(0, 0, 0, 0.1);
    min-height: 3em;
    
    &:not(:last-child) {
        margin-bottom: 1em;
    }
}


.MobileNavigationLink__link {
    text-decoration: none;
    color: $white;
    font-size: 1.5em;
    margin-top: 0.5em;
    text-align: left;
    float: left;
    margin-left: 8px;
}


.MobileNavigationLink__sublink_button {
    z-index: 2;
    float: right;
    height: 3em;
    width: calc(3em + 3px);
    margin: 0;
    padding: 0;
    color: $white;
    border: 0;
    background: transparent;
    border-left: 3px solid rgba(0, 0, 0, 0.2);


    &[data-active=false] {
        //background: rgba(0, 0, 0, 0.2);
        translate: 180deg;
    }

    &[data-active=true] {
        background: transparent;
    }
}


.MobileNavigationLink__sublink_list {
    list-style-type: none;
    width: 100%;
    overflow: hidden;

    &[data-active=false] {
        height: 0;
        color: transparent;
        background-color: transparent;
    }

    &[data-active=true] {
        height: auto;
        background-color: rgba(0, 0, 0, 0.2);
        color: $white;
    }
}


.MobileNavigationLink__sublink_link {
    float: left;
    clear: both;
    margin: 8px;

    > .MobileNavigationLink__link {
        margin-top: 0;
    }
}