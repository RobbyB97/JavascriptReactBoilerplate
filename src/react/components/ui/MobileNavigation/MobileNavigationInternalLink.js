/* Packages */
import React from 'react';
import {NavLink} from 'react-router-dom';
import {connect} from 'react-redux';


/* Actions */
import {navMenu_Mobile__Toggle} from '../../../redux/actions/ui';



/* Component */
export class MobileNavigationInternalLink extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            sublink_Menu: false
        };
    };


    sublink_Menu__Toggle = () => {
        this.setState({
            sublink_Menu: !this.state.sublink_Menu
        });
    };


    navMenu_Mobile__Toggle = () => {
        this.props.navMenu_Mobile__Toggle();
        this.setState({
            navMenu_Mobile: !this.state.navMenu_Mobile
        });
    };


    render() {
        return (
            <li className="MobileNavigationLink">
                <NavLink 
                    className="MobileNavigationLink__link"
                    to={this.props.route}
                    onClick={this.navMenu_Mobile__Toggle}
                >
                    {this.props.text}
                </NavLink>
                
                {/* Sublink menu button */}
                {this.props.sublinks &&
                    <button
                        className="MobileNavigationLink__sublink_button"
                        onClick={this.sublink_Menu__Toggle}
                        data-active={this.state.sublink_Menu}
                    >
                        â–¼
                    </button>
                }

                {/* Sublink menu */}
                {this.props.sublinks &&
                    <ul 
                        className="MobileNavigationLink__sublink_list"
                        onClick={this.sublink_Menu__Toggle}
                        data-active={this.state.sublink_Menu}
                    >
                        {this.props.sublinks.map((sublink) => (
                            <li 
                                className="MobileNavigationLink__sublink"
                                key={sublink.text}
                            >
                                <NavLink
                                    className="MobileNavigationLink__sublink_link"
                                    to={sublink.route}
                                    onClick={this.navMenu_Mobile__Toggle}
                                >
                                    {sublink.text}
                                </NavLink>
                            </li>
                        ))}
                    </ul>
                }
            </li>
        );
    };
};


/* Connect to store */
const mapDispatchToProps = (dispatch) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default connect(undefined, mapDispatchToProps)(MobileNavigationInternalLink);