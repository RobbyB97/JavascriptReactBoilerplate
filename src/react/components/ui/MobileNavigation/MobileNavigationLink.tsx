import React from 'react';
import { NavHashLink as NavLink } from 'react-router-hash-link';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { navMenu_Mobile__Toggle } from '../../../redux/actions/ui';
import { navigationLinksType } from '../../../data/navigationLinks';


interface Props extends navigationLinksType {
    navMenu_Mobile__Toggle: any;
};

interface State {
    sublink_Menu: boolean;
}


export class MobileNavigationLink extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            sublink_Menu: false
        };
    };


    sublink_Menu__Toggle: any = () => {
        this.setState({
            sublink_Menu: !this.state.sublink_Menu
        });
        
        // TODO: Close navmenu mobile when sublink clicked
    };

    navMenu_Mobile__Toggle: any = () => {
        /* Disable scrolling */
        const html = document.getElementsByTagName('HTML')[0];
        html.setAttribute('data-Mobile_Nav', 'false');

        /* Dispatch */
        this.props.navMenu_Mobile__Toggle();
    };


    render() {
        return (
            <li className="MobileNavigationLink">
                {this.props.internal == true ?
                    /* Internal link */
                    <NavLink 
                        className="MobileNavigationLink__link"
                        to={this.props.href}
                        onClick={this.navMenu_Mobile__Toggle}
                    >
                        {this.props.text}
                    </NavLink>
                    :

                    /* External link */
                    <a
                        className="MobileNavigationLink__link"
                        href={this.props.href}
                        onClick={this.navMenu_Mobile__Toggle}
                        target="_blank"
                        rel="noopener"
                    >
                        {this.props.text}
                    </a>
                }

                {/* Sublink button */}
                {this.props.sublinks &&
                    <button
                        className="MobileNavigationLink__sublink_button"
                        onClick={this.sublink_Menu__Toggle}
                        data-active={this.state.sublink_Menu}
                    >
                        â–¼
                    </button>
                }

                {/* Sublink menu */}
                {this.props.sublinks &&
                    <ul
                        className="MobileNavigationLink__sublink_list"
                        onClick={this.sublink_Menu__Toggle}
                        data-active={this.state.sublink_Menu}
                    >
                        {this.props.sublinks.map((sublink) => (
                            <li
                                className="MobileNavigationLink__sublink"
                                key={sublink.text}
                            >
                                {sublink.internal == true ?
                                    /* Internal sublink */
                                    <NavLink
                                        className="MobileNavigationLink__sublink_link"
                                        to={sublink.href}
                                        onClick={
                                            this.sublink_Menu__Toggle
                                        }
                                    >
                                        {sublink.text}
                                    </NavLink>
                                    :

                                    /* External sublink */
                                    <a
                                        className="MobileNavigationLink__sublink_link"
                                        href={sublink.href}
                                        onClick={
                                            this.sublink_Menu__Toggle
                                        }
                                    >
                                        {sublink.text}
                                    </a>
                                }
                            </li>
                        ))}
                    </ul>
                }
            </li>
        );
    };
};


/* Connect to store */
const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default connect(undefined, mapDispatchToProps)(MobileNavigationLink);