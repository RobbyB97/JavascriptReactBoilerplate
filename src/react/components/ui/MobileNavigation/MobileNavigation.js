/* Packages */
import React from 'react';
import {connect} from 'react-redux';


/* Actions */
import {navMenu_Mobile__Toggle} from '../../../redux/actions/ui';


/* Components */
import MobileNavigationInternalLink from './MobileNavigationInternalLink';
import MobileNavigationExternalLink from './MobileNavigationExternalLink';


/* Component */
export class MobileNavigation extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            navMenu_Mobile: false
        };
    };

    
    navMenu_Mobile__Toggle = () => {
        /* Disable/enable scrolling */
        const html = document.getElementsByTagName('HTML')[0];
        if (this.props.navMenu_Mobile) {
            html.setAttribute('data-Mobile_Nav', 'false');
        } else {
            html.setAttribute('data-Mobile_Nav', 'true');
        }
        
        /* Set state */
        this.props.navMenu_Mobile__Toggle();
        this.setState({
            navMenu_Mobile: !this.state.navMenu_Mobile
        });
    };


    render() {
        return (
            <section id="MobileNavigation">
                <div 
                    className="MobileNavigation__button"
                    onClick={this.navMenu_Mobile__Toggle}    
                >
                    {this.props.navMenu_Mobile ? "✕" : "☰"}
                </div>

                <nav 
                    data-active={this.props.navMenu_Mobile}
                    className="MobileNavigation__menu"
                >
                    <ul className="MobileNavigation__list">
                        {this.props.internal_links &&
                            this.props.internal_links.map((link) => (
                                <MobileNavigationInternalLink {...link}/>
                            ))
                        }

                        {this.props.external_links &&
                            this.props.external_links.map((link) => (
                                <MobileNavigationExternalLink {...link}/>
                            ))
                        }
                    </ul>
                </nav>
            </section>
        );
    };
};


/* Connect to store */
const mapStateToProps = (state) => {
    return {
        navMenu_Mobile: state.ui.navMenu_Mobile
    };
};


const mapDispatchToProps = (dispatch) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default connect(mapStateToProps, mapDispatchToProps)(MobileNavigation);