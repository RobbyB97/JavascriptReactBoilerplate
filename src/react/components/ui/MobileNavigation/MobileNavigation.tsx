import React from 'react';
import {connect} from 'react-redux';

import MobileNavigationLink from './MobileNavigationLink';
import {navMenu_Mobile__Toggle} from '../../../redux/actions/ui';
import {navigationLinksType} from '../../../data/navigationLinks'


interface Props {
    navMenu_Mobile: boolean;
    navMenu_Mobile__Toggle: any;
    links: navigationLinksType[];
}

interface State {
    navMenu_Mobile: boolean;
}


export class MobileNavigation extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            navMenu_Mobile: false
        };
    };

    
    navMenu_Mobile__Toggle = () => {
        /* Disable/enable scrolling */
        const html = document.getElementsByTagName('HTML')[0];
        if (this.props.navMenu_Mobile) {
            html.setAttribute('data-mobile_nav', 'false');
        } else {
            html.setAttribute('data-mobile_nav', 'true');
        }
        
        /* Set state */
        this.props.navMenu_Mobile__Toggle();
        this.setState({
            navMenu_Mobile: !this.state.navMenu_Mobile
        });
    };


    render() {
        return (
            <section id="MobileNavigation">
                <div 
                    className="MobileNavigation__button"
                    onClick={this.navMenu_Mobile__Toggle}
                    data-navmenu_mobile={this.props.navMenu_Mobile}  
                >
                    {this.props.navMenu_Mobile ? "✕" : "☰"}
                </div>

                <nav 
                    data-active={this.props.navMenu_Mobile}
                    className="MobileNavigation__menu"
                >
                    <ul 
                        className="MobileNavigation__list"
                        data-active={this.props.navMenu_Mobile}
                    >
                        {/* Links */}
                        {this.props.links &&
                            this.props.links.map((link) => (
                                <MobileNavigationLink 
                                    key={link.text}
                                    {...link}
                                />
                            ))
                        }
                    </ul>
                </nav>
            </section>
        );
    };
};


/* Connect to store */
const mapStateToProps = (state: any) => {
    return {
        navMenu_Mobile: state.ui.navMenu_Mobile
    };
};


const mapDispatchToProps = (dispatch: any) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default connect(mapStateToProps, mapDispatchToProps)(MobileNavigation);