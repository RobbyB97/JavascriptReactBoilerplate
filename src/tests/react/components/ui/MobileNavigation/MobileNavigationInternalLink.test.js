/* Packages */
import React from 'react';
import {shallow} from 'enzyme';


/* Component */
import {MobileNavigationInternalLink} from '../../../../../react/components/ui/MobileNavigation/MobileNavigationInternalLink'


/* Test config */
let component;
beforeEach(() => {
    component = shallow(<MobileNavigationInternalLink />)
});


/* Tests */
test('Render MobileNavigationInternalLink component', () => {
    expect(component).toMatchSnapshot();
})