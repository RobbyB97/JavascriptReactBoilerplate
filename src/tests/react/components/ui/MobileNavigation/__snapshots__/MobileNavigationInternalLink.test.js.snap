// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MobileNavigationInternalLink component 1`] = `
<li
  className="MobileNavigationLink"
>
  <NavLink
    activeClassName="active"
    aria-current="page"
    className="MobileNavigationLink__link"
    onClick={[Function]}
  />
</li>
`;
